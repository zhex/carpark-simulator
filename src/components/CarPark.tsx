import { Direction } from 'models/Bus';
import { IPosition } from 'models/interfaces/IPosition';
import * as React from 'react';
import styled from 'styled-components';

export interface ICarParkProps {
    rows: number;
    cols: number;
    position?: IPosition;
    faceTo?: Direction;
    className?: string;
    cellSize: number;
}

export interface ICellProps {
    size: number;
    className?: string;
}

export interface IRowProps {
    className?: string;
}

function createArray(n: number): any[] {
    return Array.from(new Array(n));
}

const Cell: React.SFC<ICellProps> = props => (
    <div className={props.className} />
);

export const StyledCell = styled(Cell)`
    position: relative;
    border-right: 1px solid #ccc;
    border-bottom: 1px solid #ccc;
    height: ${p => p.size + 'px'};
    flex: 1;
`;

const Row: React.SFC<IRowProps> = props => (
    <div className={props.className}>{props.children}</div>
);

export const StyledRow = styled(Row)`
    display: flex;
`;

const CarPark: React.SFC<ICarParkProps> = props => {
    const { faceTo, position, rows, cols, className, cellSize } = props;
    return (
        <div className={className}>
            {createArray(rows).map((row, idx) => (
                <StyledRow key={idx}>
                    {createArray(cols).map((col, colIdx) => (
                        <StyledCell key={colIdx} size={cellSize} />
                    ))}
                </StyledRow>
            ))}
            {props.children}
        </div>
    );
};

const StyledCarPark = styled(CarPark)`
    position: relative;
    width: ${p => `${p.cellSize! * p.cols}px`};
    border-top: 1px solid #ccc;
    border-left: 1px solid #ccc;
`;

export default StyledCarPark;
