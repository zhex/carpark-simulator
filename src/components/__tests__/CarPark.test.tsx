import 'jest-styled-components';
import * as React from 'react';
import { create } from 'react-test-renderer';

import CarPark, { StyledCell, StyledRow } from '../CarPark';

it('CarPark Component', () => {
    const tree = create(<CarPark cols={5} rows={5} cellSize={50} />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('width', '250px');
});

it('Cell Component', () => {
    const tree = create(<StyledCell size={50} />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('height', '50px');
    expect(tree).toHaveStyleRule('flex', '1');
});

it('Row Component', () => {
    const tree = create(<StyledRow />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('display', 'flex');
});
