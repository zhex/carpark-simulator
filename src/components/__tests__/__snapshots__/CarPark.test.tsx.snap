// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarPark Component 1`] = `
.c2 {
  position: relative;
  border-right: 1px solid #ccc;
  border-bottom: 1px solid #ccc;
  height: 50px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  position: relative;
  width: 250px;
  border-top: 1px solid #ccc;
  border-left: 1px solid #ccc;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
    <div
      className="c2"
    />
  </div>
</div>
`;

exports[`Cell Component 1`] = `
.c0 {
  position: relative;
  border-right: 1px solid #ccc;
  border-bottom: 1px solid #ccc;
  height: 50px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div
  className="c0"
/>
`;

exports[`Row Component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
/>
`;
